### PR TITLE
Added JavaDoc and tests for Property + Some changes to Property and Player

### PR DESCRIPTION
**Pull request by giaphuongd**
- Added PropertyTest
- Adjusted buildHouse from Player and addHouse from Property to add more conditions for a player when building houses (needs to have a sufficient amount of money, must own property and the whole colour set) 
- Added some functions that help determine whether a player owns a property colour set 
- Added getUtilities and getRailroads (to allow possible implementations that make the players pay according to the rollDice results instead of a fixed amount)

### CODE REVIEW

**Review by benxu3 — APPROVED** [review]
Could you change the rent loop variable to be less than 6 in Property when we create Property instances. Otherwise, thanks for writing the JavaDoc and Unit Tests.

**Review by derrcodes — APPROVED** [review]
Great job on writing tests and the JavaDoc. I agree with the changes (adding more conditions checks) since it would help remove any chance of execution under unwanted circumstances.

File: src/main/java/Entities/Player.java [hunk 1]
```diff
@@ -235,9 +305,20 @@ else if(a == b && (consecutive + 1) == 3){
      * @param property the property to build the house on
      * @param houses the number of houses to build
      */
-    public void buildHouse(Property property, int houses) {
-        if (properties.contains(property)) {
-            property.addHouse(this, houses);
+    public String buildHouse(Property property, int houses) {
```
Comment (Cindywwang) [inline src/main/java/Entities/Player.java:308]: nice edition to check the type of property before adding house and returning a message to the user.

File: src/main/java/Entities/Player.java [hunk 2]
```diff
@@ -109,6 +107,78 @@ public Player(String name, int money, boolean inJail, int jailCards, int positio
...
+     * Gets all the complete sets of properties owned by this player (sets are grouped by colours)
+     *
+     * @return an ArrayList of colours referring to the complete sets of properties owned by this player
+     */
+    public ArrayList<String> ownedPropertySets() {
+        ArrayList<String> ownedSets = new ArrayList<>();
+        HashMap<String, Integer> sets = createSetMap();
+        for (Property property : this.properties) {
+            sets.put(property.getColour(), sets.get(property.getColour()) + 1);
+            for (Map.Entry<String, Integer> colour : sets.entrySet()) {
```
Comment (Cindywwang) [inline src/main/java/Entities/Player.java:148]: Using two for loops here may result in a set that has duplicated key values, but if it only works this way we can change this later.

File: src/main/java/Entities/Property.java [hunk 3]
```diff
@@ -117,29 +172,63 @@ private int getHousesRent() {
...
+    /**
+     * Gets the number of houses on this property
+     * @return the number of houses built on this property
+     */
     public int getHouses(){
         return this.houses;
     }
 
-    public boolean addHouse(Player currentPlayer, int houses){
-        if (currentPlayer.getMoney() >= getHouseCost() * houses){
```
Comment (Cindywwang) [inline src/main/java/Entities/Property.java:137]: nice catch to check if the currentPlayer got enough money before adding houses.

File: src/main/java/Entities/Player.java [hunk 4]
```diff
@@ -235,9 +305,20 @@ else if(a == b && (consecutive + 1) == 3){
      * @param property the property to build the house on
      * @param houses the number of houses to build
      */
-    public void buildHouse(Property property, int houses) {
-        if (properties.contains(property)) {
-            property.addHouse(this, houses);
+    public String buildHouse(Property property, int houses) {
+        switch (property.addHouse(this, houses)) {
```
Comment (Cindywwang) [inline src/main/java/Entities/Player.java:309]: Good edition to check the type of the property before adding houses and good choice to use the switch keyword.

**Review by Cindywwang — APPROVED** [review]

**Pull request merged**